### PR TITLE
sparse_categorical_crossentropy #2894

### PR DESCRIPTION
**Pull request by wammy19**
Hi,

Here's an attempt at the Tensorflow frontend function sparse_categorical_crossentropy. I'm still getting my head round Hypothesis, and how to construct a test, any advice on that would be much appreciated 😄. 

In terms of the function itself, is this a step in the right direction?

### CODE REVIEW

File: ivy/functional/frontends/tensorflow/metrics.py [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+
+def sparse_categorical_crossentropy(
+    y_true,
+    y_pred,
+    from_logits=False,
+    axis=-1
+):
+    if from_logits:
+        y_pred = ivy.softmax(y_pred)
```
Comment (Daniel4078) [inline ivy/functional/frontends/tensorflow/metrics.py:11]: You may use the out argument of ivy.softmax to do inplace update of the data on y_pred directly in this case I think.

File: ivy/functional/frontends/tensorflow/metrics.py [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+    return ivy.sparse_cross_entropy(y_true, y_pred, axis=axis, epsilon=1e-7)
```
Comment (Daniel4078) [inline ivy/functional/frontends/tensorflow/metrics.py:12]: As epsilon's default value is 1e-7 in ivy.sparse_cross_entropy, so you may just skip this keyword argument.

File: ivy/functional/frontends/tensorflow/metrics.py [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+
+
+sparse_categorical_crossentropy.unsupported_dtypes = {"torch": ("float16", "bfloat16")}
```
Comment (Daniel4078) [inline ivy/functional/frontends/tensorflow/metrics.py:15]: By the way, please add a blank line at the end of files according to formatting rule. Thank you.

File: ivy_tests/test_ivy/test_frontends/test_tensorflow/test_metrics.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+        num_positional_args,
+        fw
+):
+    y_true = np.array(y_true, dtype='int32')
+    dtype = y_pred[0]
+    y_pred = np.array(y_pred[1], dtype=dtype)
+
+    # Perform softmax on prediction if it's not a probability distribution.
+    if not from_logits:
+        y_pred = np.exp(y_pred) / np.sum(np.exp(y_pred))
```
Comment (Daniel4078) [inline ivy_tests/test_ivy/test_frontends/test_tensorflow/test_metrics.py:36]: The test should only use ivy functions and does not rely on other framworks.

File: ivy_tests/test_ivy/test_frontends/test_tensorflow/test_metrics.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+)
+def test_sparse_categorical_crossentropy(
+        y_true,
+        y_pred,
+        from_logits,
+        num_positional_args,
+        fw
+):
+    y_true = np.array(y_true, dtype='int32')
+    dtype = y_pred[0]
```
Comment (Daniel4078) [inline ivy_tests/test_ivy/test_frontends/test_tensorflow/test_metrics.py:31]: you can do this instead:
dtype, y_pred = dtype_y_pred
and use dtype_y_pred to pass the result of helpers.dtype_and_values
in this way, y_pred is naturally a ivy.array and dtype can be used later for the input_dtypes argument.
Please check the docstring of the relative functions and look for examples in other test files.

File: ivy_tests/test_ivy/test_frontends/test_tensorflow/test_metrics.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+        with_out=False,
+        num_positional_args=num_positional_args,
+        native_array_flags=True,
+        fw=fw,
+        frontend="tensorflow",
+        fn_tree="metrics.sparse_categorical_crossentropy",
+        y_true=y_true,
+        y_pred=y_pred,
+        from_logits=from_logits
+    )
```
Comment (Daniel4078) [inline ivy_tests/test_ivy/test_frontends/test_tensorflow/test_metrics.py:50]: Please don't forget to add a blank line at the end of the file.

File: ivy_tests/test_ivy/test_frontends/test_tensorflow/test_metrics.py [hunk 7]
```diff
@@ -0,0 +1,50 @@
+# global
+import numpy as np
+from hypothesis import given, strategies as st
+
+# local
+import ivy_tests.test_ivy.helpers as helpers
+import ivy.functional.backends.tensorflow as ivy_tf
+
+
```
Comment (Daniel4078) [inline ivy_tests/test_ivy/test_frontends/test_tensorflow/test_metrics.py:9]: The use of the  decorator handle_cmd_line_args has been changed a little. So now we're wrapping the  given decorator, instead of the test function. This is done so as to set the backend before assigning any values in the given decorator. We should now be using ivy.current_backend_str().valid_dtypes inside our helper functions, instead of ivy_np.valid_dtypes(i'm currently in the process of updating, this). Also, please don't pass in anything for as_variable, native_array, container, with_out, instance_method inside given, since that is being handled by the decorator, and it will only overwrite the values for these flags, which we don't want.

**Review by Daniel4078 — CHANGES_REQUESTED** [review]
Thank you for your contribution and there are some changes to be done before your code can be merged. Please check the comments below. If you have any questions, please ask.

**Comment by Daniel4078** [comment]
just a small tip, you can request a fresh review from a reviewer by click the refresh icon in the Reviewers section of the sidebar of the Conversation tab.

File: ivy_tests/test_ivy/test_frontends/test_tensorflow/test_metrics.py (hunk 7 above)
Comment (wammy19) [reply ivy_tests/test_ivy/test_frontends/test_tensorflow/test_metrics.py:10]: I've made some of the changes you mentioned here but I don't think I'm fully understanding everything.

"We should now be using ivy.current_backend_str().valid_dtypes inside our helper functions, instead of ivy_np.valid_dtypes"

`ivy.current_backend_str()` this function returns `Union[str, None]` so therefore it cannot have an attribute of `.valid_dtypes`. Did you mean to mention another function else? Also is this to replace the use of `ivy_tf.valid_float_dtypes` on line 16?

The test itself is failing still, but I'm not totally sure why, I've look at recently merged tests as examples and can't seem to figure out what's wrong, perhaps I'm misusing the `@handle_cmd_line_args` function wrapper?

Thank you for your help 🙃.

**Comment by Daniel4078** [comment]
firstly, the tf_tree argument in the test should hold the path to the function in frontend framework namespace, in this case it should be keras.metrics.sparse_categorical_crossentropy.

**Review by Daniel4078 — CHANGES_REQUESTED** [review]
Currently, there is no formatting error in the code that you wrote (excluding the changed you merged from the main branch). But two errors are exposed by the test you wrote.
Firstly, the tf_tree argument in the test should hold the path to the function in frontend framework namespace, in this case it should be keras.metrics.sparse_categorical_crossentropy.
Secondly, TensorFlow does not support inplace updates of the tf.Tensor, so I actually gave a bad advice as you should not use the ivy.softmax function's out argument to do inplace update but do it as how you did it before.
Thank you!

File: ivy_tests/test_ivy/test_frontends/test_tensorflow/test_metrics.py (hunk 7 above)
Comment (Daniel4078) [reply ivy_tests/test_ivy/test_frontends/test_tensorflow/test_metrics.py:10]: Well, I supposed it should be about ivy.current_backend() funcition

**Comment by wammy19** [comment]
Ok awesome, these changes seem to make the test pass.

**Review by Daniel4078 — CHANGES_REQUESTED** [review]
Well, the checks just show that you have a logical error in the test you wrote. To be more specific, the class labels the function takes are indexed starting from 0, so as you set the number of classes to be 5, then the class label should be 0-4, not 0-5. Also, do not forget to add one blank line at the end of each file you write. Thank you.

**Comment by wammy19** [comment]
> Well, the checks just show that you have a logical error in the test you wrote. To be more specific, the labels you generate are indexed starting from 0, so as you set the number of classes to be 5, then the class label should be 0-4, not 1-5.Also, do not forget to add one blank line at the end of each file you write. Thank you.

Yeah I do actually have a blank line on my end in pycharm, ended up adding two blank lines to make it appear here on github. Not too sure why 🤷‍♂️.

**Comment by wammy19** [comment]
I can see that the files I just pushed don't have a blank line at the end. going to add a third blank line on my end and see if that fixes it.

**Review by Daniel4078 — APPROVED** [review]
Now I think there is nothing wrong here. You may try remove all other irrelevant changes from this pull request with only the changes on the two files you are working on left. In this way, you can greatly reduce the risk of merging outdated code to the main branch that someone just fixed.
